Linux/BSD: Fix support for NetBSD

akien-mga opened the pull request:
Add __NetBSD__ to `platform_config.h` so that it can find `alloca`
and use the proper `pthread_setname_np` format.

Rename RANDOM_MAX to avoid conflict with NetBSD stdlib.

Fixes #42145.

@notepade Can you check if this works for you, and if there are other issues?

---

NetBSD: Implement OS_Unix::get_executable_path()

Same implementation as OpenBSD seems to work fine.

---

X11: Try to load libXrandr.so.3 if libXrandr.so.2 isn't found

All Linux distros, and FreeBSD and OpenBSD seem to have libXrandr.so.2,
but for some reason recent NetBSD versions seem to have libXrandr.so.3 now.

akien-mga committed:
Linux/BSD: Fix support for NetBSD

Add __NetBSD__ to `platform_config.h` so that it can find `alloca`
and use the proper `pthread_setname_np` format.

Rename RANDOM_MAX to avoid conflict with NetBSD stdlib.

Fixes #42145.

aaronfranke commented on core/math/math_funcs.h:
```diff
@@ -46,7 +46,8 @@ class Math {
 
-	static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
+	// Not using 'RANDOM_MAX' to avoid conflict with system headers on some OSes (at least NetBSD).
+	static const uint64_t RANDOM_UINT32_MAX = 0xFFFFFFFF;
```
> `UINT32`

> `uint64_t`

Hmm...

akien-mga replied:
It's the max value of a 32-bit unsigned integer, stored in `uint64_t`. Is there a problem?

aaronfranke replied:
Is there a particular reason this shouldn't be either `uint32_t` or `0xFFFFFFFFFFFFFFFF`?

Anyway, the name seems a bit strange and some may mistake it for Hungarian notation. Maybe `RANDOM_32BIT_MAX` would be clearer?

akien-mga replied:
> Is there a particular reason this shouldn't be either `uint32_t` or `0xFFFFFFFFFFFFFFFF`?

No idea.

> Anyway, the name seems a bit strange and some may mistake it for Hungarian notation. Maybe `RANDOM_32BIT_MAX` would be clearer?

Sounds fine.

akien-mga commented:
There are more fixes needed for NetBSD support, marking as WIP for now.

akien-mga commented:
Cherry-picked version for `3.2` (which I'm using for testing for now): https://github.com/akien-mga/godot/tree/3.2-netbsd

akien-mga committed:
NetBSD: Implement OS_Unix::get_executable_path()

Same implementation as OpenBSD seems to work fine.

akien-mga committed:
X11: Try to load libXrandr.so.3 if libXrandr.so.2 isn't found

All Linux distros, and FreeBSD and OpenBSD seem to have libXrandr.so.2,
but for some reason recent NetBSD versions seem to have libXrandr.so.3 now.

akien-mga commented:
The first commit solves the compilation errors (#42145).

The other two commits solve runtime errors about missing libXrandr.so.2 (for some reason NetBSD has libXrandr.so.3) and unimplemented `OS::get_executable_path()` (the OpenBSD code seems to work fine for NetBSD too).

So here's Dodge the Creeps running on NetBSD 9.0 in VirtualBox with the llvmpipe software rasterizer.
(Don't mind the cropped window, I couldn't get VBox additions to work so I'm stuck with 800x600.)

![Screenshot_20200918_120316](https://user-images.githubusercontent.com/4701338/93585448-2000cf00-f9a7-11ea-9492-c779e2ac5bfd.png)

akien-mga commented:
For the reference, I'm pushing this to `master`, but the current `master` branch can't be built on NetBSD with those changes yet, as the vendored vulkan-loader library doesn't support it.

There's a WIP NetBSD package for vulkan-loader which could likely be used instead of the vendored lib once it's ready for production: https://pkgsrc.se/wip/vulkan-loader

But that's not a priority for me at this point, we can revisit this later when there's out-of-the-box Vulkan support in NetBSD. (I haven't checked if there are any GPU drivers with Vulkan support for NetBSD currently.)

The `3.2` version of this PR does make Godot 3.2.3+ functional on NetBSD: https://github.com/akien-mga/godot/tree/3.2-netbsd

akien-mga commented:
Cherry-picked for 3.2.4.

notepade commented:
> For the reference, I'm pushing this to `master`, but the current `master` branch can't be built on NetBSD with those changes yet, as the vendored vulkan-loader library doesn't support it.
> 
> There's a WIP NetBSD package for vulkan-loader which could likely be used instead of the vendored lib once it's ready for production: https://pkgsrc.se/wip/vulkan-loader
> 
> But that's not a priority for me at this point, we can revisit this later when there's out-of-the-box Vulkan support in NetBSD. (I haven't checked if there are any GPU drivers with Vulkan support for NetBSD currently.)
> 
> The `3.2` version of this PR does make Godot 3.2.3+ functional on NetBSD: https://github.com/akien-mga/godot/tree/3.2-netbsd

I'm a bit confused what you mean by all this. You're saying that it won't be compilable on NetBSD even if you pushed that pull request?

akien-mga commented:
The latest `3.2` branch can be compiled on NetBSD now. The `master` branch can up to a certain point where it fails due to the lack of Vulkan support, which is beyond the scope of this PR.

niacat commented:
Can libXrandr be loaded unversioned? The library will have a different version number on NetBSD depending on the way X11 is installed, libXrandr.so.3 if the base system's X11 sets are installed, libXrandr.so.2 if installed as modular X11 through the packaging system.

akien-mga commented:
> Can libXrandr be loaded unversioned? 

No, because unversioned library symlinks (`libXrandr.so`) are part of devel packages on Linux, and not installed for most regular users.

One thing that could be looked into though is using Xrandr directly without dlopen'ing it, as we seem to link it anyway for other code. This dlopen maybe dates back to before we had a build dependency on Xrandr.

niacat commented:
Ah, OK. On every BSD there is no concept of separation of dev packages, so it should be fine to attempt to load the version provided by upstream Xorg (libXrandr.so.2), then fall back to unversioned.

The history of this is apparently that we were previously including XFree86 as part of the OS, then as part of the switchover to X.org a decade ago every library got its version bumped. So, I think it's very unlikely the version number will change in the future and am not too worried.
